#160452435: Create login page

uchemukolo opened the pull request:
#### What does this PR do?
- Display login modal
- Added validations on the Login form
- Consumed backend API for login an existing user

#### Description of Task to be completed?
- Add unit tests, integration tests and e2e tests
- Display alert to the user when an error occurs during login
- Close the modal box on successful login.
- Redirect to Homepage after successful login
- implement log out functionality


#### How should this be manually tested?
- Fetch and checkout to the branch
```$xslt
> git fetch
> git checkout ft-160452435-user-login
> npm install 
```
- Rename the `.env.sample` file on the root directory of the cloned repo to `.env`. Uncomment and supply the following variable: `REACT_APP_HOST` which is the link to the frontend application; `REACT_APP_BASE_API_URL` which is the base URL for the backend api. For example, `https://odin-ah-backend-staging.herokuapp.com/api/v1`.
- To run the test first start the frontend server.

```$xslt 
> npm start
```
then run
```$xslt 
> npm run test
```
- Open http://localhost:3000 in your browser and click on the `Login `
**Note**: Ensure to start the backend api server if `REACT_APP_BASE_API_URL` is pointing to your local machine.
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
- [#160452435](https://www.pivotaltracker.com/story/show/160452435)
#### Screenshots (if appropriate)
![image](https://user-images.githubusercontent.com/30410066/48339349-4d241e00-e668-11e8-8516-a0a7f456244e.png)


#### Questions:

waleCloud reviewed (commented):
Increase test coverage, endeavor to cover all necessary files created/modified. before raising a final PR.
well done so far! 👍

macphilips commented on src/components/Header/HeaderContainer.jsx:
```diff
@@ -0,0 +1,85 @@
+    this.setState(state => ({ showLoginModal: !state.showLoginModal }));
+  };
+
+  handleLoginCloseClick = () => {
```
Hi @uchemukolo, I noticed the functions on line 41 and 45 are the both doing the same thing. You just one of the function to change the state of `showLoginModal`. I would suggest you remove one of the function and rename the other function (e.g. `toggleLoginModalState`).

uchemukolo replied:
Thank you @macphilips. Feedback noted and implemented

macphilips commented on src/services/apiRequest.js:
```diff
@@ -0,0 +1,74 @@
+      (error) => {
+        const { status } = error.response;
+        if (status === 401) {
+          // todo clear authentication token from local storage
```
Great job the task: clearing authentication token from local storage. Since you are done implementing the task you can remove the comments. Also, I noticed from your `src/rudex/actions/login.js` file, you implemented a function for logging out a user.
```js
export const logout = () => {
  localStorage.removeItem('jwtToken');
  return {
    type: LOGOUT_USER
  };
};
``` 
 I think it is better to dispatch the `logout` action here rather than the `loginUserFailure` and remove `localStorage.clear();` since you are also removing the token in the `logout` function.

uchemukolo replied:
Feedback noted and implemented

macphilips commented on src/tests/actions/LoginAction.test.js:
```diff
@@ -0,0 +1,139 @@
+// import axios from 'axios';
```
Kindly remove this commented code

uchemukolo replied:
Removed

Commit:
feature(UI): Create login page

- Add unit tests for the components, actions and reducers created
- Design login modal
- Added validations on Login form
- Consumed backend api for login an existing user
- implement log out functionality

[Delivers #160452435]

macphilips commented on src/components/home/Homee.js:
```diff
@@ -0,0 +1,47 @@
+import React, { Component } from 'react';
```
Hi @uchemukolo, I noticed this file is similar to the `src/components/home/Home.js`, kindly address this duplicate

uchemukolo replied:
Yes. the duplicate has been deleted and e2e test added.

macphilips reviewed (commented):
You did a good job on this task. I noticed you added some unit and integration tests, well done. However, I did not find any end-2-end (e2e) test on this PR. Kindly add some e2e tests and we are good to go. Lastly, update the PR description (the section `Task to be completed` to be specific) and update screenshot too. Thanks

macphilips reviewed (approved):
PR looks good to me. It meets all acceptance criteria for this feature. I approved this PR to be merged.
